app/vmagent/remotewrite/remotewrite.go

```diff
@@ -748,16 +748,27 @@ func (rwctx *remoteWriteCtx) reinitStreamAggr() {
 		logger.Errorf("cannot reload stream aggregation config from -remoteWrite.streamAggr.config=%q; continue using the previously loaded config; error: %s", sasFile, err)
 		return
 	}
-	if !sasNew.Equal(sas) {
-		sasOld := rwctx.sas.Swap(sasNew)
-		sasOld.MustStop()
-		logger.Infof("successfully reloaded stream aggregation configs at -remoteWrite.streamAggr.config=%q", sasFile)
-	} else {
+
+	defer func() {
+		metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_streamaggr_config_reload_successful{path=%q}`, sasFile)).Set(1)
+		metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_streamaggr_config_reload_success_timestamp_seconds{path=%q}`, sasFile)).Set(fasttime.UnixTimestamp())
+	}()
+
+	if sasNew.Equal(sas) {
 		sasNew.MustStop()
 		logger.Infof("the config at -remoteWrite.streamAggr.config=%q wasn't changed", sasFile)
+		return
 	}
-	metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_streamaggr_config_reload_successful{path=%q}`, sasFile)).Set(1)
-	metrics.GetOrCreateCounter(fmt.Sprintf(`vmagent_streamaggr_config_reload_success_timestamp_seconds{path=%q}`, sasFile)).Set(fasttime.UnixTimestamp())
+
+	sasOld := rwctx.sas.Load()
+	sasOldLen := sasOld.Len()
+	updated := sasOld.UpdateWith(sasNew)
```

Comment:
I think that you change here the value of rwctx.sas by the pointer and next line do nothing, because pointer isn't changed. Probably that this code may not be thread-safe.

Reply:
>  because pointer isn't changed

why do you think so? `func (a *Aggregators) UpdateWith(b *Aggregators) int` seem to operate on pointer receiver

Reply:
@Amper please see update in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4862/commits/e45a1247a4b6d8b9011cb198ff937561fd0e04b7